pkg/model/coordinator/coordinator.go

```diff
@@ -210,35 +209,53 @@ func (coo *Coordinator) InitState(bootstrap bool, startIndex milestone.Index) er
 	}
 	cachedBndl.Release()
 
-	coo.lastCheckpointHash = &(coo.state.LatestMilestoneHash)
+	coo.lastCheckpointHash = coo.state.LatestMilestoneHash
+	coo.lastMilestoneHash = coo.state.LatestMilestoneHash
 	coo.bootstrapped = true
 	return nil
 }
 
-// issueCheckpoint sends a secret checkpoint transaction to the network.
-// we do that to prevent parasite chain attacks.
-// only honest tipselection will reference our checkpoints, so the milestone will reference honest tips.
-func (coo *Coordinator) issueCheckpoint() error {
+// issueCheckpointWithoutLocking tries to create and send a "checkpoint" to the network.
+// a checkpoint can contain multiple chained transactions to reference big parts of the unconfirmed cone.
+// this is done to keep the confirmation rate as high as possible, even if there is an attack ongoing.
+// new checkpoints always reference the last checkpoint or the last milestone if it is the first checkpoint after a new milestone.
+func (coo *Coordinator) issueCheckpointWithoutLocking(enforceTips bool) error {
 
-	tip, err := coo.tipselFunc()
-	if err != nil {
-		return err
+	if !tangle.IsNodeSynced() {
+		return tangle.ErrNodeNotSynced
 	}
 
-	b, err := createCheckpoint(tip, *coo.lastCheckpointHash, coo.minWeightMagnitude, coo.powFunc)
+	tips, err := coo.checkpointTipselFunc(enforceTips)
 	if err != nil {
 		return err
 	}
 
-	if err := coo.sendBundleFunc(b); err != nil {
-		return err
+	var lastCheckpointHash hornet.Hash
+	if coo.lastCheckpointIndex == 0 {
+		// reference the last milestone
+		lastCheckpointHash = coo.lastMilestoneHash
+	} else {
+		// reference the last checkpoint
+		lastCheckpointHash = coo.lastCheckpointHash
 	}
 
-	coo.lastCheckpointCount++
-	lastCheckpointHash := hornet.HashFromHashTrytes(b[0].Hash)
-	coo.lastCheckpointHash = &lastCheckpointHash
+	for i, tip := range tips {
```

Comment:
This issues n checkpoints for n tips. In a tree the same result could be achieved with only log(n) checkpoints. Especially since the checkpoint require a considerable amount of PoW this should be considered!